packages/block-editor/src/hooks/background.js

```diff
@@ -302,18 +340,150 @@ function BackgroundImagePanelItem( { clientId, setAttributes } ) {
 	);
 }
 
+function backgroundSizeHelpText( value ) {
+	if ( value === 'cover' || value === undefined ) {
+		return __( 'Stretch image to cover the block.' );
+	}
+	if ( value === 'contain' ) {
+		return __( 'Image is contained within the block.' );
+	}
+	return __( 'Repeat the image, and set a fixed size.' );
+}
+
+function BackgroundSizePanelItem( {
+	clientId,
+	isShownByDefault,
+	setAttributes,
+} ) {
+	const style = useSelect(
+		( select ) =>
+			select( blockEditorStore ).getBlockAttributes( clientId )?.style,
+		[ clientId ]
+	);
+
+	const value = style?.background?.backgroundSize;
+
+	const hasValue = hasBackgroundSizeValue( style );
+
+	const resetAllFilter = useCallback( ( previousValue ) => {
+		return {
+			...previousValue,
+			style: {
+				...previousValue.style,
+				background: {
+					...previousValue.style?.background,
+					backgroundSize: undefined,
+				},
+			},
+		};
+	}, [] );
+
+	const updateBackgroundSize = ( next ) => {
+		setAttributes( {
+			style: cleanEmptyObject( {
+				...style,
+				background: {
+					...style?.background,
+					backgroundSize: next,
+				},
+			} ),
+		} );
+	};
+
+	// An `undefined` value is treated as `cover` by the toggle group control.
+	// An empty string is treated as `auto` by the toggle group control. This
+	// allows a user to select "Size" and then enter a custom value, with an
+	// empty value being treated as `auto`.
+	const currentValueForToggle =
+		( value !== undefined && value !== 'cover' && value !== 'contain' ) ||
+		value === ''
+			? 'auto'
+			: value || 'cover';
+
+	return (
+		<VStack
+			as={ ToolsPanelItem }
+			spacing={ 2 }
+			className="single-column"
+			hasValue={ () => hasValue }
+			label={ __( 'Background size' ) }
+			onDeselect={ () => resetBackgroundSize( style, setAttributes ) }
+			isShownByDefault={ isShownByDefault }
+			resetAllFilter={ resetAllFilter }
+			panelId={ clientId }
+		>
+			<ToggleGroupControl
+				__nextHasNoMarginBottom
+				size={ '__unstable-large' }
+				label={ __( 'Background size' ) }
+				value={ currentValueForToggle }
+				onChange={ updateBackgroundSize }
+				isBlock={ true }
+				help={ backgroundSizeHelpText( value ) }
+			>
+				<ToggleGroupControlOption
+					key={ 'cover' }
+					value={ 'cover' }
+					label={ __( 'Cover' ) }
+				/>
+				<ToggleGroupControlOption
+					key={ 'contain' }
+					value={ 'contain' }
+					label={ __( 'Contain' ) }
+				/>
+				<ToggleGroupControlOption
+					key={ 'size' }
+					value={ 'auto' }
+					label={ __( 'Size' ) }
+				/>
+			</ToggleGroupControl>
+			{ value !== undefined &&
+			value !== 'cover' &&
+			value !== 'contain' ? (
```

Comment:
Is there a reason not to use `value === 'size'`?

Reply:
Yes, we never actually set the value to `size` (size is just the current label and key for the control). In order for the active `ToggleGroupControl` to match the expected value, we need to do some mapping here between what the real value is, and what the group control thinks the value is. So the logic is:

* Any value that will either be output as `auto` or a fixed size, we treat as `auto` so that the Size toggle is highlighted. That toggle button's value needs to be `auto` so that clicking on sets the correct value.
* We then fall back to `value` to be used to determine which control is active.
* Finally, sue the default fallback of `cover` in case anything has slipped through the conditions (this is mostly so that `null` is treated as `cover`), though it shouldn't _really_ ever get hit.

TL;DR, we use 'size' as a label, but `auto` or a fixed value as the real value, and need to treat that as `auto` for the right button to be highlighted.

Hope that all makes sense, let me know if you think it can be simplified, though!

Reply:
Thanks for the explainer!